// engines/test/shared/helpers.js
var fs = require('fs');
var readFile = fs.readFile;
var readFileSync = fs.readFileSync;
var handlebars = require('handlebars');
var engines = require('../../');


exports.test = function (name) {
  var user = {
    name: '<strong>Assemble</strong>'
  };

  describe(name, function () {

    afterEach(function () {
      fs.readFile = readFile;
      fs.readFileSync = readFileSync;
    });

    if (name == 'handlebars') {
      // Use case: return safe HTML that won’t be escaped in the final render.
      it('should support helpers', function (done) {
        var str = fs.readFileSync('test/fixtures/' + name + '/helpers.' + name).toString();

        var locals = {
          user: user,
          helpers: {
            safe: function (object) {
              return new handlebars.SafeString(object);
            }
          }
        };

        engines[name].render(str, locals, function (err, html) {
          if (err) return done(err);
          html.should.equal('<strong>Assemble</strong>');
          done();
        });
      });
    }
  });
};
