// engines/test/shared/partials.js
var fs = require('fs');
var readFile = fs.readFile;
var readFileSync = fs.readFileSync;
var engines = require('../../');



exports.test = function (name) {
  var user = {
    name: 'Assemble'
  };


  describe(name, function () {
    afterEach(function () {
      fs.readFile = readFile;
      fs.readFileSync = readFileSync;
    });


    if (name == 'hogan' || name == 'mustache' || name == 'handlebars' || name == 'ractive') {
      it('should support partials', function (done) {
        var path = 'test/fixtures/' + name + '/partials.' + name;
        var locals = {
          user: user,
          partials: {
            partial: 'user'
          }
        };

        engines[name](path, locals, function (err, html) {
          if (err) {return done(err);}
          html.should.equal('<p>Assemble</p>');
          done();
        });
      });
    } else {
      it('should support rendering a partial', function (done) {
        var str = fs.readFileSync('test/fixtures/' + name + '/user_partial.' + name).toString();
        var locals = {
          user: user,
          views: "./test/fixtures/" + name
        };

        engines[name].render(str, locals, function (err, html) {
          if (err) {return done(err);}
          html.should.equal('<p>Assemble from partial!</p><p>Assemble</p>');
          done();
        });
      });
    }
  });
};
